### PR TITLE
Refactoring storeCodeFromRoute, resolve store code on SSR

### LINKED ISSUE #3576: storeCodeFromRoute route object without host breaks loop
## Current behavior
In `core/lib/multistore.ts::storeCodeFromRoute` if `matchedRouteOrUrl` is an object but is missing the `host` property, the current logic simply returns an empty string if the first store in the store loop does not match. The logic is basically the following:

1. Iterate all stores from `mapStoreUrlsFor`
2. In each iteration, create `normalizedPath` from `matchedRouteOrUrl.path` (if object)
3. If `normalizedPath` doesn't match the store code in the current iteration, check if `matchedRouteOrUrl` is an object. 
4. If it IS an object, and it doesn't have `['host']` then return `''`.

The problem here is only the first store is checked under these conditions. The specific scenario where I discovered this is when I am on a product view and I try to load it server side, it fails and redirects to page not found. I'm not sure why the route object doesn't have `host`, but this is what's causing this break in the loop.

## Expected behavior
It should follow the same logic as when `matchedRouteOrUrl` is a string. So instead of `return ''`, it should either `continue`, or set `url = normalizedPath`.

## Steps to reproduce the issue
I'm not sure if there's something in our environment that is causing this scenario to work this way, but how we produced it is:
1. Enable multistore with at least two stores active. 
2. Go to a product view on a store that is 2nd or later in the list of `mapStoreUrlsFor`. 
3. Refresh the page.

## Can you handle fixing this bug by yourself?

- [x] YES
- [ ] NO

## Which [Release Cycle](https://docs.vuestorefront.io/guide/basics/release-cycle.html) state this refers to? Info for developer.
Pick one option.

- [x] This is a bug report for test version on https://test.storefrontcloud.io - In this case Developer should create branch from `develop` branch and create Pull Request `2. Feature / Improvement` back to `develop`.
- [ ] This is a bug report for current Release Candidate version on https://next.storefrontcloud.io - In this case Developer should create branch from `release` branch and create Pull Request `3. Stabilisation fix` back to `release`.
- [ ] This is a bug report for current Stable version on https://demo.storefrontcloud.io and should be placed in next stable version hotfix - In this case Developer should create branch from `hotfix` or `master` branch and create Pull Request `4. Hotfix` back to `hotfix`.

## Environment details

- Node: 10.16
- Code Version: v1.10.2

### PR DESCRIPTION
**Pull request by andrzejewsky**
### Related issues
closes #3576

### Short description and why it's useful
It solves the problem when we want to reveal store code on SSR.

### Which environment this relates to
Check your case. In case of any doubts please read about [Release Cycle](https://docs.vuestorefront.io/guide/basics/release-cycle.html)

- [x] Test version (https://test.storefrontcloud.io) - this is a new feature or improvement for Vue Storefront. I've created branch from `develop` branch and want to merge it back to `develop`
- [x] RC version (https://next.storefrontcloud.io) - this is a stabilisation fix for Release Candidate of Vue Storefront. I've created branch from `release` branch and want to merge it back to `release`
- [ ] Stable version (https://demo.storefrontcloud.io) - this is an important fix for current stable version. I've created branch from `hotfix` or `master` branch and want to merge it back to `hotfix`

### Upgrade Notes and Changelog

- [x] No upgrade steps required (100% backward compatibility and no breaking changes)
- [ ] I've updated the [Upgrade notes](https://github.com/DivanteLtd/vue-storefront/blob/develop/docs/guide/upgrade-notes/README.md) and [Changelog](https://github.com/DivanteLtd/vue-storefront/blob/develop/CHANGELOG.md) on how to port existing VS sites with this new feature

### Contribution and currently important rules acceptance
- [x] I read and followed [contribution rules](https://github.com/DivanteLtd/vue-storefront/blob/master/CONTRIBUTING.md)

### CODE REVIEW

**Comment by andrzejewsky** [comment]
@pkarw I'm just started with reviewing this PR, it probably will be a conflict as we have modified same file (but not so complicated) so would be nice to merge it beforehand.

**Comment by pkarw** [comment]
Cool; @patzick was into CR of #3540 as well so please just sync with him and merge

File: core/lib/storeCodeFromRoute.ts [hunk 1]
```diff
@@ -0,0 +1,49 @@
+import { RawLocation } from 'vue-router'
+import config from 'config'
+import { LocalizedRoute } from './types'
+
+const getNormalizedPath = (matchedRouteOrUrl) => {
+  const matchingPath = typeof matchedRouteOrUrl === 'object' ? matchedRouteOrUrl.path : matchedRouteOrUrl
```
Comment (patzick) [inline core/lib/storeCodeFromRoute.ts]: we're not prepared here for null
```suggestion
  const matchingPath = matchedRouteOrUrl && matchedRouteOrUrl.path || matchedRouteOrUrl
```

File: core/lib/storeCodeFromRoute.ts [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+
+  return matchingPath.length > 0 && matchingPath[0] !== '/' ? `/${matchingPath}` : matchingPath
```
Comment (patzick) [inline core/lib/storeCodeFromRoute.ts]:
```suggestion
  return matchingPath && matchingPath.length > 0 && matchingPath[0] !== '/' ? `/${matchingPath}` : matchingPath
```

would be nice to have these methods unit tested

File: core/lib/storeCodeFromRoute.ts [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+}
+
+const getUrl = (matchedRouteOrUrl) => {
+  const normalizedPath = getNormalizedPath(matchedRouteOrUrl)
+
+  if (typeof matchedRouteOrUrl === 'object') {
```
Comment (patzick) [inline core/lib/storeCodeFromRoute.ts]: here also we have no cover for null

**Pull request merged**